walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,159 @@
+      TimeRange nextElem = busyTimes.get(next);
+      endPoint = nextElem.overlaps(meetingTime) ? next : next - 1;
+      endTime = nextElem.overlaps(meetingTime) ? nextElem.end() : meetingTime.end();
+    }
+
+    // create a new meeting with the given start and end times
+    TimeRange newBusy = TimeRange.fromStartEnd(startTime, endTime, false);
+
+    /*
+     * This meeting ends before the current first meeting starts, so add it to the
+     * beginning of the list
+     */
+    if (endPoint < 0) {
+      newBusyTimes.add(newBusy);
+    }
+    for (int i = 0; i < busyTimes.size(); i++) {
+      if (i == startPoint && endPoint >= 0 && startPoint < busyTimes.size()) {
+        // Replace all the entries between startPoint and endPoint with the coalesced entry
+        newBusyTimes.add(newBusy);
+        i = endPoint;
+      } else {
+        newBusyTimes.add(busyTimes.get(i));
+      }
+    }
+    /*
+     * This meeting starts after the current last meeting ends, so add it to the end of
+     * the list
+     */
+    if (startPoint >= busyTimes.size() && endPoint >= 0) {
+      newBusyTimes.add(newBusy);
+    }
+    return newBusyTimes;
+  }
+
+  /*
+   * Given a list of busy times in non-overlapping, sorted order and a meeting duration
+   * returns a list of time slots of atleast 'duration' length during which no meetings
+   * are scheduled
+   */
+  private ArrayList<TimeRange> findFreeTimes(ArrayList<TimeRange> busyTimes, long duration) {
+    Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
+    ArrayList<TimeRange> freeTimes = new ArrayList<>();
+    int start = TimeRange.START_OF_DAY;
+    int end = TimeRange.END_OF_DAY;
+    for (int i = 0; i < busyTimes.size(); i++) {
+      TimeRange curr = busyTimes.get(i);
+
+      int thisStart = curr.start();
+      int thisEnd = curr.end();
+      /*
+       * create a free slot from the beginning of the free period to 'thisStart',
+       * the beginning of the next busy one
+       */
+      TimeRange newFree = TimeRange.fromStartEnd(start, thisStart, false);
+      if (newFree.duration() >= duration) {
+        freeTimes.add(newFree);
+      }
+      start = thisEnd;
+    }
+    // Create a free slot from the end of the last meeting to the end of the day if possible
+    TimeRange newFree = TimeRange.fromStartEnd(start, end, true);
+    if (newFree.duration() >= duration) {
+      freeTimes.add(newFree);
+    }
+    return freeTimes;
+  }
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> busy = new ArrayList<>();
+    HashSet<String> attendees = new HashSet<>(request.getAttendees());
+    Iterator<Event> iterator = events.iterator();
+    while (iterator.hasNext()) {
+      Event meeting = iterator.next();
+      Set<String> meetingAttendees = meeting.getAttendees();
+      if (Sets.intersection(attendees, meetingAttendees).isEmpty()) {
```

Comment:
Instead of `if (condition) { continue; } else { do something }` you can just do `if (not condition) { do something }`